### ISSUE TITLE
[vmagent] keepalive not working properly

### ISSUE THREAD

**Issue by posix** [VictoriaMetrics/VictoriaMetrics#663]
**Describe the bug**
It looks like something is wrong with keepalive, since when checking network connections via netstat, I see  frequently reconnects to the remote datastore and the keepalive state is only visible on the remote socket.
As example of a stable keepalive operation is telegraf - ports do not change and reconnects do not occur.
There is also a possibility that data is lost because of this, for example: we write data into two databases and sometimes it is noticeable that one has data, and the other does not(short interval from 1 to 5 seconds ).

**To Reproduce**
Run vmagent with remote storage and check connectin throgh **netstat -alnpo**.

**Screenshots**
![vma2vm_2](https://user-images.githubusercontent.com/20062930/89063327-7f534300-d357-11ea-8ffe-95b92774a888.gif)
![vma2vm](https://user-images.githubusercontent.com/20062930/89063334-837f6080-d357-11ea-987f-a11024ffedc7.gif)
![vma2vma](https://user-images.githubusercontent.com/20062930/89063336-84b08d80-d357-11ea-990e-e3190993c6b2.gif)

**Version**
```
vmagent-prod --version
vmagent-20200731-091128-heads-master-0-g3f498cf2

 victoria-metrics-prod --version
victoria-metrics-20200730-215334-heads-master-0-g44a86e1b
```

**Used command-line flags**
```
cat /etc/default/vmagent2
VMA_OPTS='-fs.disableMmap -http.disableResponseCompression -httpListenAddr=:8189 -remoteWrite.tmpDataPath /data/vm/agent2 -memory.allowedPercent 20 -remoteWrite.maxDiskUsagePerURL 0 -loggerLevel ERROR -remoteWrite.url http://127.0.0.1:8428/api/v1/write'
```
```
VMA_OPTS='-maxConcurrentInserts 256 -fs.disableMmap -http.disableResponseCompression -httpListenAddr=":8428" -remoteWrite.queues 4 -remoteWrite.tmpDataPath /data/vm/agent -memory.allowedPercent 30 -remoteWrite.maxDiskUsagePerURL 0 -loggerLevel ERROR -promscrape.config /data/vm/etc/vma.yml -remoteWrite.url http://127.0.0.1:8429/api/v1/write -remoteWrite.url http://s4877:8429/api/v1/write'
```
```
cat /etc/default/victoriametrics
VM_OPTS='-search.minStalenessInterval 1s -fs.disableMmap -http.disableResponseCompression -storageDataPath /data/vm/dt -maxConcurrentInserts 512 -memory.allowedPercent 80 -retentionPeriod 5 -search.maxQueryDuration 120s -httpListenAddr=:8429 -loggerLevel ERROR'
```

**Additional context**
I tried many things to fix this: changed sysctl network settings, changed local and remote configurations to vmagent and vma, and got no useful result.

**Comment by valyala** [comment]
@posix , could you try `vmagent` v1.38.1? There were significant changes in how vmagent sends data to remote storage between v1.38.1 and v1.39.0 in the commit ad6290953cec458b891c9effddf2210b2ca90dd2 , which could lead to broken keep-alive connections

**Comment by valyala** [comment]
Could you also set up monitoring for `vmagent` according to [these docs](https://victoriametrics.github.io/vmagent.html#monitoring) and then share graphs for `rate(vmagent_remotewrite_dials_total[1m])` query? This query shows the rate for new connections to remote storage. It would be great if this graph will be shown for `vmagent` v1.38.1 and for `vmagent` v1.39.2 .

**Comment by posix** [comment]
> @posix , could you try `vmagent` v1.38.1? There were significant changes in how vmagent sends data to remote storage between v1.38.1 and v1.39.0 in the commit [ad62909](https://github.com/VictoriaMetrics/VictoriaMetrics/commit/ad6290953cec458b891c9effddf2210b2ca90dd2) , which could lead to broken keep-alive connections
Yes, i tried this version and i ve got same result:
![prod_vma_1 39 2](https://user-images.githubusercontent.com/20062930/89313518-1d9d1c80-d668-11ea-9055-e9138cd3c365.gif)
![prod_vma_1 38 1](https://user-images.githubusercontent.com/20062930/89313532-22fa6700-d668-11ea-8b7c-a9265a757a29.gif)
And strange situation on our test nodes:
![test_vma_1 39 2](https://user-images.githubusercontent.com/20062930/89313602-3c031800-d668-11ea-9ed2-30fe88a09c97.gif)
![test_vma_1 38 1](https://user-images.githubusercontent.com/20062930/89313617-41606280-d668-11ea-9afc-0c1ff6ab8f25.gif)

**Comment by valyala** [comment]
@posix , could you check for error logs in `vmagent` and `victoriametrics` when connections from `vmagent` to `victoriametrics` are re-established?

**Comment by posix** [comment]
> Could you also set up monitoring for `vmagent` according to [these docs](https://victoriametrics.github.io/vmagent.html#monitoring) and then share graphs for `rate(vmagent_remotewrite_dials_total[1m])` query? This query shows the rate for new connections to remote storage. It would be great if this graph will be shown for `vmagent` v1.38.1 and for `vmagent` v1.39.2 .

![ka_1 39 2](https://user-images.githubusercontent.com/20062930/89326483-e5063e80-d679-11ea-9cb9-c799566e6aaf.png)

![ka_rate_1 38](https://user-images.githubusercontent.com/20062930/89326703-43332180-d67a-11ea-9bf2-442447ba743c.png)

**Comment by posix** [comment]
> @posix , could you check for error logs in `vmagent` and `victoriametrics` when connections from `vmagent` to `victoriametrics` are re-established?
There are no errors in the logs.

**Comment by valyala** [comment]
Could you try vmagent from the commit c25b0c2cd581ddada0189d51d709b443a5c87cca ?

**Comment by posix** [comment]
> Could you try vmagent from the commit [c25b0c2](https://github.com/VictoriaMetrics/VictoriaMetrics/commit/c25b0c2cd581ddada0189d51d709b443a5c87cca) ?

It looks better and there are no reconnections, but keepalive is still "off".
![Screenshot 2020-08-04 at 21 10 02](https://user-images.githubusercontent.com/20062930/89329276-1ed94400-d67e-11ea-850d-e50cb30b69b0.png)
![prod_vma_1 39 2](https://user-images.githubusercontent.com/20062930/89329308-28fb4280-d67e-11ea-800d-2df2e701b9a4.gif)

**Comment by valyala** [comment]
The `keepalive` shown by `netstat` is related to [TCP keepalive mechanism](https://en.wikipedia.org/wiki/Keepalive#TCP_keepalive), which allows detecting broken connection when the remote side is no longer available. TCP keepalive may be useful for connections with very infrequent data transfers (i.e. a single data transfer per hour). It isn't useful in general for connections with relatively frequent data transfers such as connections from `vmagent` to `-remoteWrite.url`, since the broken connection will be determined as soon as the next data transfer fails.

As for [HTTP keepalive](https://en.wikipedia.org/wiki/HTTP_persistent_connection), which allows re-using the same TCP connection for sending multiple HTTP requests, it is enabled by default on both `vmagent` and `VictoriaMetrics` sides. This is confirmed by `rate(vmagent_remotewrite_dials_total[1m])` query results - it returns the per-second rate of establishing new connections from `vmagent` to remote storage. If it is close to zero while `vmagent` sends data to remote storage, then this means that connections are re-used.

**Comment by valyala** [comment]
FYI, the bugfix that prevents from creating new connection per request to remote storage has been included in [v1.39.3](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing this issue as fixed.

@posix , feel free re-opening the issue if you still see that `rate(vmagent_remotewrite_dials_total[1m])` returns big numbers for `vmagent` v1.39.3 or higher. It should return values close to zero under normal conditions.
